protocols/noise/src/io/framed.rs

```diff
@@ -0,0 +1,429 @@
+    Eof,
+    /// An encryption error occurred (terminal state).
+    EncErr
+}
+
+impl WriteState {
+    fn is_ready(&self) -> bool {
+        if let WriteState::Ready = self {
+            return true
+        }
+        false
+    }
+}
+
+impl<T, S> futures::stream::Stream for NoiseFramed<T, S>
+where
+    T: AsyncRead + Unpin,
+    S: SessionState + Unpin
+{
+    type Item = io::Result<Vec<u8>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let mut this = Pin::into_inner(self);
+        loop {
+            trace!("read state: {:?}", this.read_state);
+            match this.read_state {
+                ReadState::Ready => {
+                    this.read_state = ReadState::ReadLen { buf: [0, 0], off: 0 };
+                }
+                ReadState::ReadLen { mut buf, mut off } => {
+                    let n = match read_frame_len(&mut this.io, cx, &mut buf, &mut off) {
+                        Poll::Ready(Ok(Some(n))) => n,
+                        Poll::Ready(Ok(None)) => {
+                            trace!("read: eof");
+                            this.read_state = ReadState::Eof(Ok(()));
+                            return Poll::Ready(None)
+                        }
+                        Poll::Ready(Err(e)) => {
+                            return Poll::Ready(Some(Err(e)))
+                        }
+                        Poll::Pending => {
+                            this.read_state = ReadState::ReadLen { buf, off };
+                            return Poll::Pending;
+                        }
+                    };
+                    trace!("read: frame len = {}", n);
+                    if n == 0 {
+                        trace!("read: empty frame");
+                        this.read_state = ReadState::Ready;
+                        continue
+                    }
+                    this.read_buffer.resize(usize::from(n), 0u8);
+                    this.read_state = ReadState::ReadData { len: usize::from(n), off: 0 }
+                }
+                ReadState::ReadData { len, ref mut off } => {
+                    let n = {
+                        let f = Pin::new(&mut this.io).poll_read(cx, &mut this.read_buffer[*off .. len]);
+                        match ready!(f) {
+                            Ok(n) => n,
+                            Err(e) => return Poll::Ready(Some(Err(e))),
+                        }
+                    };
+                    trace!("read: {}/{} bytes", *off + n, len);
+                    if n == 0 {
+                        trace!("read: eof");
+                        this.read_state = ReadState::Eof(Err(()));
+                        return Poll::Ready(Some(Err(io::ErrorKind::UnexpectedEof.into())))
+                    }
+                    *off += n;
+                    if len == *off {
+                        trace!("read: decrypting {} bytes", len);
+                        let mut frame = vec![0u8; len];
```

Comment:
This does introduce a new buffer contrary to the PR's description, no? Previously we reused our internal buffers and copied the data to/from the provided `Async{Read,Write}` argument slices. Now we create a fresh decryption buffer on every message.

Reply:
That is true, it essentially comes with the separation of framed (`NoiseFramed`) and streaming (`NoiseOutput`) I/O. At this point when `NoiseFramed` is wrapped in the `NoiseOutput` we would ideally like to decrypt directly into the frame buffer of `NoiseOutput` instead of exchanging the current frame buffer with the new frame. Though I don't know how to avoid this or what compiler optimisations may make of it. Do you have an idea for a different approach? The fundamental problem is that using `NoiseOutput` during the handshake phase is no longer desired, because it "hides" the noise framing already during the handshake phase. Separating these two stages seemed like the obvious route to take and as a small positive side-effect removed some code duplication between the `AsyncWrite::poll_write` and `AsyncWrite::poll_flush` of `NoiseOutput`.

Reply:
> Though I don't know how to avoid this or what compiler optimisations may make of it.

I do not think the compiler can optimise away the vector allocations.

> Do you have an idea for a different approach? The fundamental problem is that using `NoiseOutput` during the handshake phase is no longer desired, because it "hides" the noise framing already during the handshake phase. 

`NoiseOutput` could have inherent private methods — thus visible to the child module `handshake` — that support reading and returning the frame length which the handshake state uses.

Reply:
What do you think of https://github.com/libp2p/rust-libp2p/pull/1658/commits/6a4fc139ea1a81e83332e11cadbae4b7c6bad9b6? `bytes` seems to be cut out for this use-case.

Reply:
Using `BytesMut` is better, but I am not sure if we really need to split `NoiseOutput` and add another `NoiseFramed` type. For reference this is what I had in mind: https://github.com/libp2p/rust-libp2p/compare/master...twittner:noise-read-frame.

Reply:
Thanks. I made the split not only to address the problem though, but at the same time saw an opportunity to improve the code - two birds with one stone so to speak. I'm sad to hear if you don't share that view. But before I discard this approach, let me try to motivate more clearly why I perceive the split as an improvement:

  * It removes some code duplication in `AsyncWrite::poll_write` and `AsyncWrite::poll_flush` by design, i.e. turns [these two blocks](https://github.com/libp2p/rust-libp2p/blob/master/protocols/noise/src/io.rs#L237-L395) into one in the `Sink` impl of `NoiseFramed`.
  * The structure of the code reflects more clearly the structure of the I/O, i.e. the noise framing on top of which an encrypted stream is provided more clearly stands out and is reflected in the `NoiseOutput(NoiseFramed)` layering, instead of being "built-in" to `AsyncRead` / `AsyncWrite`. I.e. one larger block of code is split into two simpler ones, each a bit easier to understand on its own.
  * In a similar vain, the handshake phase and the transport phase can explicitly operate on distinct I/O types, using `NoiseFramed` with the convenient `Stream` / `Sink` API in handshake mode and wrapping it in the encrypted `NoiseOutput` `AsyncRead` / `AsyncWrite` stream when changing to transport mode.

In your diff, there is `NoiseOutput::read_frame`, returning the frame length, which is used in tandem with the `AsyncRead` API to write frame-aware code during the handshake, and writing frames remains by means of `poll_write` + `flush` (knowing that this will send exactly one frame). I still perceive this like a workaround for letting one peek into the abstraction over the noise framing done by the `AsyncRead` + `AsyncWrite` impl, indicating (to me) that this abstraction comes too early and should be applied only when the handshake is complete, using a frame-based API during the handshake.

If this is still unconvincing and you don't see the changes overall as an improvement (or even as a step backward), I would somewhat reluctantly give in. Let me know.

Reply:
> * It removes some code duplication in `AsyncWrite::poll_write` and `AsyncWrite::poll_flush` by design, i.e. turns [these two blocks](https://github.com/libp2p/rust-libp2p/blob/master/protocols/noise/src/io.rs#L237-L395) into one in the `Sink` impl of `NoiseFramed`.

I did not look at the writing part at all. Would it not be possible to use an inherent method for `NoiseOutput` which is shared between writing and flushing?

> I made the split not only to address the problem though, but at the same time saw an opportunity to improve the code - two birds with one stone so to speak. I'm sad to hear if you don't share that view.  [...] If this is still unconvincing and you don't see the changes overall as an improvement (or even as a step backward), I would somewhat reluctantly give in and follow the road shown in your diff (or maybe you want to submit it as a PR?). Let me know.

The primary complication to me is the explicit state machine handling which is still there. Adding inherent methods to provide access to frames does not seem that different to me compared to doing so via separate `Sink`/`Stream` impls which then require `BytesMut` to mitigate the overhead that comes with it. But regardless, the code I linked to was merely an illustration and not meant to replace this PR.

Reply:
> I did not look at the writing part at all. Would it not be possible to use an inherent method for NoiseOutput which is shared between writing and flushing?

Maybe, I didn't try because that part "came naturally" with the split of `NoiseOutput` and `NoiseFramed`. It wasn't really my intention to make refactorings in this PR, like removing duplicated code, that are not in direct connection to the changes for frame-based handshake I/O.

> The primary complication to me is the explicit state machine handling which is still there.

True, but I think it is an improvement nevertheless.

> Adding inherent methods to provide access to frames does not seem that different to me compared to doing so via separate Sink/Stream impls [..]

I take it then that you really see no value in actually having clearly different types and APIs for handshake and transport mode and don't share my feeling that combinations of `read_frame`, `poll_write`, `poll_read` and `flush` to read and write frames during handshake mode, instead of inherently frame-based `send` and `next`, seem like an unclean approach to writing frame-aware code during the handshake phase. :white_flag:.

> [..] which then require BytesMut to mitigate the overhead that comes with it.

I don't think of using `BytesMut` as an unwelcome requirement to avoid presumed overhead that would come with an implementation that does not use it, but rather as an inherent part of a good implementation that does such a split. However, I only ran some unscientific "benchmarks" using the existing smoke tests with various fixed payload sizes, in which these changes even came out a bit on top but at least showed no regression.

> But regardless, the code I linked to was merely an illustration and not meant to replace this PR

Since the approaches are very different, I don't see much point in adapting this PR, however. So closing it in favor of a new one seems like the best way to go, regardless of who does it.

Reply:
> [...] presumed overhead [...]

I do hope that Rust allows us to spot allocations from reading source code and therefore *conclude* rather than *presume* if there is overhead or not ;-)

> Since the approaches are very different, I don't see much point in adapting this PR, however. So closing it in favor of a new one seems like the best way to go, regardless of who does it.

Why close it? As I wrote above: "the code I linked to was [...] not meant to replace this PR".